File: servers/axum/src/streaming.rs
// probatorium axum adapter — WS/SSE streaming surface.
//
// Mirrors the gorilla_ws reference (servers/gorilla_ws/server.go) so the
// celeris streaming cell-columns have a like-for-like tower-stack baseline.
// Two routes, both HTTP/1.1 only (the WS upgrade and the SSE long-poll both
// ride h1; the h2c serve path never exercises them):
//
//   GET /ws      — RFC6455 upgrade; ?mode= selects echo / large-echo / hub.
//   GET /events  — SSE fan-out; every subscriber streams the published frame.
//
// Fan-out is a single broadcast "tick": one task pulses `()` every 1ms and
// every hub socket / SSE subscriber emits its own frame on each tick. This
// replaces the gorilla reference's per-broker connection-set + RWMutex with
// tokio's broadcast channel — slow consumers see `Lagged` and skip, which is
// exactly the drop-rather-than-block policy the reference's non-blocking
// channel send implements.

use std::collections::HashMap;
use std::time::Duration;

use axum::extract::ws::{Message, WebSocket, WebSocketUpgrade};
use axum::extract::{Query, State};
use axum::response::sse::{Event, KeepAlive, Sse};
use axum::response::Response;
use futures_util::stream::{self, Stream, StreamExt};
use tokio::sync::broadcast;

// ?mode= selectors — mirror scenarios.StreamMode* and the gorilla reference
// so the loadgen client drives every adapter identically.
const STREAM_MODE_WS_HUB: &str = "ws-hub";
const STREAM_MODE_WS_LARGE_ECHO: &str = "ws-large-echo";

const BROADCAST_PAYLOAD: &str = "payload";
const SSE_EVENT_DATA: &str = "hello";

const STREAM_PUBLISH_INTERVAL: Duration = Duration::from_millis(1);

// 1 MiB read limit for the large-echo mode — holds the 64 KiB large-echo
// payload with headroom, matching the gorilla reference's SetReadLimit.
const LARGE_ECHO_READ_LIMIT: usize = 1 << 20;

// Tick-channel capacity. A subscriber that falls this far behind the 1ms
// publisher sees Lagged and skips ahead — the drop policy. Sized generously
// so only genuinely stalled consumers ever lag.
const TICK_CHANNEL_CAPACITY: usize = 1024;

// AppState carries the shared broadcast tick. Cloning a Sender is cheap
// (Arc bump) and every handler subscribes off it, so the static handlers
// stay state-free while the streaming handlers extract this.
#[derive(Clone)]
pub struct AppState {
    tick: broadcast::Sender<()>,
}

impl AppState {
    pub fn new() -> Self {
        let (tick, _) = broadcast::channel(TICK_CHANNEL_CAPACITY);
        Self { tick }
    }

    // spawn_publisher starts the single 1ms pulse that drives every hub
    // socket and SSE subscriber. send() errors only when there are zero
    // receivers; that is the steady idle state, so the error is discarded.
    pub fn spawn_publisher(&self) {
        let tick = self.tick.clone();
        tokio::spawn(async move {
            let mut interval = tokio::time::interval(STREAM_PUBLISH_INTERVAL);
            loop {
                interval.tick().await;
                let _ = tick.send(());
            }
        });
    }
}

// ws_handler upgrades the connection and dispatches on ?mode=. large-echo
// raises the per-message read limit before the upgrade is negotiated.
pub async fn ws_handler(
    ws: WebSocketUpgrade,
    Query(params): Query<HashMap<String, String>>,
    State(state): State<AppState>,
) -> Response {
    let mode = params.get("mode").cloned();
    match mode.as_deref() {
        Some(STREAM_MODE_WS_HUB) => {
            let rx = state.tick.subscribe();
            ws.on_upgrade(move |socket| ws_hub(socket, rx))
        }
        Some(STREAM_MODE_WS_LARGE_ECHO) => ws
            .max_message_size(LARGE_ECHO_READ_LIMIT)
            .max_frame_size(LARGE_ECHO_READ_LIMIT)
            .on_upgrade(ws_echo),
        _ => ws.on_upgrade(ws_echo),
    }
}

// ws_echo reflects every received frame back verbatim (same opcode). Any
// read or write error ends the loop and drops the socket.
async fn ws_echo(mut socket: WebSocket) {
    while let Some(Ok(msg)) = socket.recv().await {
        // Stop on control-close; only data frames are echoed. recv() already
        // answers Ping with Pong internally, so we just forward the rest.
        if matches!(msg, Message::Close(_)) {
            return;
        }
        if socket.send(msg).await.is_err() {
            return;
        }
    }
}

// ws_hub is a pure broadcast RECEIVER: it never echoes. On each tick it
// pushes the TEXT "payload" frame, while concurrently draining inbound
// frames (control frames, eventual close) so the peer's reads are serviced
// and a close is noticed promptly. Lagged ticks are skipped (drop policy).
async fn ws_hub(mut socket: WebSocket, mut rx: broadcast::Receiver<()>) {
    loop {
        tokio::select! {
            tick = rx.recv() => match tick {
                Ok(()) => {
                    if socket.send(Message::text(BROADCAST_PAYLOAD)).await.is_err() {
                        return;
                    }
                }
                Err(broadcast::error::RecvError::Lagged(_)) => continue,
                Err(broadcast::error::RecvError::Closed) => return,
            },
            inbound = socket.recv() => match inbound {
                Some(Ok(_)) => {} // drain; receivers never echo
                _ => return,      // read error or peer gone
            },
        }
    }
}

// sse_handler answers 200 with the event-stream content type and streams a
// `data: hello\n\n` frame on every tick. A lagging subscriber drops the
// missed ticks (filter_map skips Lagged) rather than back-pressuring the
// publisher. The stream ends when the client disconnects: axum drops the
// response future, which drops the receiver.
pub async fn sse_handler(
    State(state): State<AppState>,
) -> Sse<impl Stream<Item = Result<Event, std::convert::Infallible>>> {
    let rx = state.tick.subscribe();
    let stream = stream::unfold(rx, |mut rx| async move {
        loop {
            match rx.recv().await {
                Ok(()) => {
                    let event = Event::default().data(SSE_EVENT_DATA);
                    return Some((Ok(event), rx));
                }
                Err(broadcast::error::RecvError::Lagged(_)) => continue,
                Err(broadcast::error::RecvError::Closed) => return None,
            }
        }
    });
    // KeepAlive is harmless here (the 1ms publisher already keeps the
    // connection hot) but matches axum's idiomatic SSE setup and guards the
    // idle window before the first subscriber tick.
    Sse::new(stream.boxed()).keep_alive(KeepAlive::default())
}

File: servers/hono/src/streaming.ts
// WS/SSE streaming surface for the Bun (h1) path — the like-for-like
// counterpart to the gorilla_ws reference adapter (servers/gorilla_ws).
//
// Two routes ride the existing Bun.serve HTTP/1.1 listener:
//   - GET /ws      WebSocket; ?mode= selects echo / large-echo / hub-receiver
//   - GET /events  SSE fan-out; every subscriber joins the broker, a
//                  background publisher pushes a steady event stream
//
// Both are H1-only by design (the WS upgrade + SSE long-poll ride HTTP/1.1);
// this module is never wired into the h2c (node:http2) branch.

import type { ServerWebSocket, Server, WebSocketHandler } from "bun";

// ?mode= selectors — mirror the gorilla_ws reference + scenarios.StreamMode*
// so the loadgen client drives every adapter identically.
const STREAM_MODE_WS_HUB = "ws-hub";
const STREAM_MODE_WS_LARGE_ECHO = "ws-large-echo";

// 1 MiB read limit for large-echo; holds the 64 KiB large-echo payload.
export const LARGE_ECHO_MAX_PAYLOAD = 1 << 20;

const BROADCAST_PAYLOAD = "payload";
const SSE_EVENT_DATA = "hello";
const STREAM_PUBLISH_INTERVAL_MS = 1;

// Per-socket data threaded through server.upgrade → websocket handlers; the
// mode chosen at upgrade time decides receiver-vs-echo behaviour in message().
interface WSData {
  mode: string;
}

type HonoWebSocket = ServerWebSocket<WSData>;

// --- WebSocket hub (the naive connection-set the celeris Hub replaces) ---

// Pure broadcast receivers; the ticker fans BROADCAST_PAYLOAD to each. Echo
// sockets never join — they reply inline in message() and are not iterated.
const hub = new Set<HonoWebSocket>();

// Bun's native websocket handler. open() decides registration by mode;
// message() echoes for the echo modes; close() always leaves the hub.
export const websocket: WebSocketHandler<WSData> = {
  maxPayloadLength: LARGE_ECHO_MAX_PAYLOAD,
  open(ws: HonoWebSocket): void {
    if (ws.data.mode === STREAM_MODE_WS_HUB) hub.add(ws);
  },
  message(ws: HonoWebSocket, message: string | Buffer): void {
    // Hub receivers drain only; everything else echoes verbatim.
    if (ws.data.mode === STREAM_MODE_WS_HUB) return;
    ws.send(message);
  },
  close(ws: HonoWebSocket): void {
    hub.delete(ws);
  },
};

// --- SSE broker (the publish-to-N-subscribers fan-out the celeris Broker replaces) ---

// Pre-encode the SSE wire frame once: "data: hello\n\n".
const SSE_FRAME = new TextEncoder().encode(`data: ${SSE_EVENT_DATA}\n\n`);

// Subscribed response controllers. A slow/closed controller is dropped on the
// failing enqueue rather than blocking the shared publish loop.
const subscribers = new Set<ReadableStreamDefaultController<Uint8Array>>();

// --- background tickers ---

let broadcastTimer: ReturnType<typeof setInterval> | undefined;

// startBroadcast launches the single 1ms loop that drives both fan-outs: the
// WS hub broadcast and the SSE publish share one tick so the wire cadence
// matches the reference's two independent tickers without doubling timers.
export function startBroadcast(): void {
  if (broadcastTimer !== undefined) return;
  broadcastTimer = setInterval(() => {
    for (const ws of hub) ws.send(BROADCAST_PAYLOAD);
    for (const controller of subscribers) {
      try {
        controller.enqueue(SSE_FRAME);
      } catch {
        // closed or backpressured controller: drop rather than block the loop.
        subscribers.delete(controller);
      }
    }
  }, STREAM_PUBLISH_INTERVAL_MS);
}

export function stopBroadcast(): void {
  if (broadcastTimer !== undefined) {
    clearInterval(broadcastTimer);
    broadcastTimer = undefined;
  }
}

// --- fetch interception ---

const SSE_HEADERS: Record<string, string> = {
  "Content-Type": "text/event-stream",
  "Cache-Control": "no-cache",
  Connection: "keep-alive",
};

// handleStreaming intercepts the two streaming routes. Returns a Response for
// /events, undefined for a successful /ws upgrade (Bun owns the socket), and
// null for everything else so the caller falls through to app.fetch.
export function handleStreaming(
  req: Request,
  server: Server<WSData>,
): Response | undefined | null {
  const url = new URL(req.url);

  if (url.pathname === "/ws") {
    const mode = url.searchParams.get("mode") ?? "";
    // upgrade() hands the socket to the websocket handler on success; returning
    // undefined tells Bun the response is owned by the upgraded connection.
    if (server.upgrade(req, { data: { mode } })) return undefined;
    return new Response("expected websocket upgrade", { status: 426 });
  }

  if (url.pathname === "/events") {
    // Capture the controller so cancel() can deregister this exact subscriber
    // when the client disconnects (the explicit-unsubscribe path); the publish
    // loop's drop-on-enqueue is the backstop for controllers that error first.
    let ctrl: ReadableStreamDefaultController<Uint8Array> | undefined;
    const stream = new ReadableStream<Uint8Array>({
      start(controller): void {
        ctrl = controller;
        subscribers.add(controller);
      },
      cancel(): void {
        if (ctrl !== undefined) subscribers.delete(ctrl);
      },
    });
    return new Response(stream, { headers: SSE_HEADERS });
  }

  return null;
}
